File: Editorial/0057_Insert_Interval/Go/2_Binary_Search.go
func insert(intervals [][]int, newInterval []int) [][]int {
    // If the intervals vector is empty, return a vector containing the newInterval
    if len(intervals) == 0 {
        return [][]int{newInterval}
    }

    n := len(intervals)
    target := newInterval[0]
    left, right := 0, n-1

    // Binary search to find the position to insert newInterval
    for left <= right {
        mid := (left + right) / 2
        if intervals[mid][0] < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    // Insert newInterval at the found position
    result := make([][]int, 0)
    result = append(result, intervals[:left]...)
    result = append(result, newInterval)
    result = append(result, intervals[left:]...)

    // Merge overlapping intervals
    res := make([][]int, 0)
    for _, interval := range result {
        // If res is empty or there is no overlap, add the interval to the result
        if len(res) == 0 || res[len(res)-1][1] < interval[0] {
            res = append(res, interval)
            // If there is an overlap, merge the intervals by updating the end of the last interval in res
        } else {
            if res[len(res)-1][1] < interval[1] {
                res[len(res)-1][1] = interval[1]
            }
        }
    }

    return res
}

File: Editorial/2701-2800/2774_Array_Upper_Bound/TypeScript/2_Binary_Search.ts
declare global {
    interface Array<T> {
        upperBound(target: number): number;
    }
}

Array.prototype.upperBound = function(target: number): number {
    let left = 0;
    let right = this.length - 1;
    let result = -1; // Target not found

    while (left <= right) {
        const mid = left + ((right - left) >> 1);

        if (this[mid] === target) {
            result = mid;
            left = mid + 1; // Continue searching in the right half for the last occurrence
        } else if (this[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return result;
};
